lib/compat/wordpress-6.5/fonts/fonts.php

```diff
@@ -85,31 +85,147 @@ function gutenberg_create_initial_post_types() {
+	if ( in_array( 'do_not_allow', $caps, true ) ) {
+		// It's already known that the user is not allowed to perform the requested capability.
+		return $caps;
+	}
+
+	if ( 'delete_post' !== $cap ) {
+		// This filter is only concerned with the 'delete_post' meta capability.
+		return $caps;
+	}
+
+	$post = get_post( $args[0] );
+	if ( ! $post ) {
+		// Do not allow deleting posts that do not exist.
+		$caps[] = 'do_not_allow';
+		return $caps;
+	}
+
+	// Check for font post types.
+	$post_type = get_post_type( $post );
+	if ( 'wp_font_face' === $post_type ) {
+		// Are there any font files associated with this font face?
+		$font_files = get_post_meta( $post->ID, '_wp_font_face_file', false );
+		if ( empty( $font_files ) ) {
+			/*
+			* No font files.
+			*
+			* The user can delete the post based on the 'delete_post' meta capability.
+			*/
+			return $caps;
+		}
+
+		// The user can only delete the post if they can modify the file system.
+		$caps[] = 'upload_fonts';
+		return $caps;
+	}
+
+	if ( 'wp_font_family' === $post_type ) {
+		// Are there any font faces associated with this font family?
+		$font_faces = get_children(
+			array(
+				'post_parent' => $post->ID,
+				'post_type'   => 'wp_font_face',
+			)
+		);
+
+		if ( empty( $font_faces ) ) {
+			/*
+			 * No font faces.
+			 *
+			 * The user can delete the post based on the 'delete_post' meta capability.
+			 */
+			return $caps;
+		}
+
+		// If any of the font faces contain files, the user needs to be able to modify the file system.
+		foreach ( $font_faces as $font_face ) {
+			$font_files = get_post_meta( $font_face->ID, '_wp_font_face_file', false );
+			if ( ! empty( $font_files ) ) {
+				$caps[] = 'upload_fonts';
+				// File system caps are required, so no need to check further.
+				break;
+			}
+		}
+		return $caps;
+	}
+
+	// Return existing caps if the post type is not a font family or font face.
+	return $caps;
+}
+add_filter( 'map_meta_cap', 'gutenberg_delete_font_post_meta_caps', 10, 4 );
+
+/**
+ * Filters the user capabilities to grant the 'upload_fonts' capability as necessary.
+ *
+ * To grant the 'upload_fonts' capability, file modifications must be allowed, the fonts directory must be
+ * writable, and the user must have the 'edit_theme_options' capability.
+ *
+ * @since 6.5.0
+ *
+ * @param  bool[] $allcaps An array of all the user's capabilities.
+ * @return bool[] Filtered array of the user's capabilities.
+ */
+function gutenberg_maybe_grant_upload_font_cap( $allcaps, $caps ) {
+	if ( ! in_array( 'upload_fonts', $caps, true ) ) {
+		return $allcaps;
+	}
+
+	$fonts_dir = wp_get_font_dir()['path'];
+	$post_type = get_post_type_object( 'wp_font_face' );
+	if (
+		wp_is_file_mod_allowed( 'can_upload_fonts' ) &&
+		wp_is_writable( $fonts_dir ) &&
+		current_user_can( $post_type->cap->create_posts )
+	) {
+		$allcaps['upload_fonts'] = true;
+	}
+
+	return $allcaps;
+}
+add_filter( 'user_has_cap', 'gutenberg_maybe_grant_upload_font_cap', 10, 2 );
```

Comment:
Technically speaking, the `user_has_cap` filter's recommended use is to grant capabilities _generally_, based on other capabilities.

For specific checks like `wp_is_file_mod_allowed()` and `wp_is_writable()`, it is recommended to use the `map_meta_cap` filter instead.

In other words, I'd suggest:
* Simplify this function to only focus on granting `upload_fonts` to any user that can `$post_type->cap->create_posts`.
* In a separate `map_meta_cap` filter callback, include `do_not_allow` as a required capability if the relevant file mods aren't allowed or if the directory is not writable. This way it will effectively not be granted to anyone, but the actual capability grant logic is still tied to purely the other capabilities on the relevant user.

Reply:
> focus on granting upload_fonts to any user that can $post_type->cap->create_posts

Hmm, afaik "contributors" can create posts (that they cannot publish), but cannot upload files. Not sure if they should be able to upload (and install) fonts, seems inconsistent? 

Also don't think "authors" should be able to upload and install fonts. The reason is: how are they going to be able to use a font after it has been uploaded? May be missing something but afaik "authors" cannot make CSS changes, neither "global" nor "local". So what's the point in uploading a file you cannot use?

There are the reasons why I'm thinking a new `upload_fonts` capability looks mostly redundant. It should match exactly `edit_theme_options` as far as I understand it, and should not be used separately. I can see a need for it to simplify capability checks if `DISALLOW_FILE_MODS` would affect font files, but this use would be somewhat inconsistent. I mean, users that have `upload_fonts` will also have be able to use these newly uploaded fonts or the UI and UX will be a total mess :)

Reply:
I think your response fits much better on the discussion in the issue rather than my feedback here. My comment here was specifically about the code and is not related to the larger conversation we're having.
Could you put this on the issue as well please so we can discuss in a single place?